### ISSUE TITLE
Test suite doesn't handle some multiples in YAML files

### ISSUE THREAD

**Issue by tunetheweb** [sqlfluff/sqlfluff#1273]
I noticed this in https://github.com/sqlfluff/sqlfluff/pull/1269#discussion_r687560810 but also stumbled across it again in a more recent PR.

To repeat edit the `test/fixtures/parser/bigquery/select_mixture_of_array_literals.sql` file and change this line:

```sql
    ['a'] AS strcol1,
```

to this:

```sql
    ['a', 'b'] AS strcol1,
```

Still valid SQL right?

We also change the equivalent YAML file from this:

```yml
            array_literal:
              start_square_bracket: '['
              expression:
                literal: "'a'"
              end_square_bracket: ']'
```

to this:

```yml
            array_literal:
              start_square_bracket: '['
              expression:
                literal: "'a'"
              comma: ','
              expression:
                literal: "'b'"
              end_square_bracket: ']'
```

We then run the test suite:

```
tox -e cov-init,py38,cov-report,linting
```

The SQL test completes fine (it's valid SQL after all):

```
test/dialects/dialects_test.py::test__dialect__base_file_parse[bigquery-select_mixture_of_array_literals.sql] ✓                                                                            35% ███▌      
 ```

However the YAML check later fails:

```py
――――――――――――――――――――――――――――――――――――――― test__dialect__base_parse_struct[bigquery-select_mixture_of_array_literals.sql-True-select_mixture_of_array_literals.yml] ――――――――――――――――――――――――――――――――――――――――

dialect = 'bigquery', sqlfile = 'select_mixture_of_array_literals.sql', code_only = True, yamlfile = 'select_mixture_of_array_literals.yml', yaml_loader = <function load_yaml at 0x7feb35a62940>

    @pytest.mark.parametrize("dialect,sqlfile,code_only,yamlfile", parse_structure_examples)
    def test__dialect__base_parse_struct(
        dialect, sqlfile, code_only, yamlfile, yaml_loader
    ):
        """For given test examples, check parsed structure against yaml."""
        # Load the right dialect
        config = FluffConfig(overrides=dict(dialect=dialect))
        # Load the SQL
        raw = load_file(dialect, sqlfile)
        # Lex and parse the file
        tokens, _ = Lexer(config=config).lex(raw)
        parsed = Parser(config=config).parse(tokens)
        # Load the YAML
        res = yaml_loader(make_dialect_path(dialect, yamlfile))
        if parsed:
>           assert parsed.to_tuple(code_only=code_only, show_raw=True) == res

...
E                        (('expression',
E                          (('array_literal',
E                            (('start_square_bracket',
E                              '['),
E                             ('expression',
E                              (('literal',
E             -                  "'b'"),)),
E             ?                    ^
E             +                  "'a'"),)),
E             ?                    ^
E                             ('comma',
E                              ','),
E             +               ('expression',
E             +                (('literal',
E             +                  "'b'"),)),
E                             ('end_square_bracket',
E                              ']'))),)),

...

 test/dialects/dialects_test.py::test__dialect__base_parse_struct[bigquery-select_mixture_of_array_literals.sql-True-select_mixture_of_array_literals.yml] ⨯                                59% █████▉    
 ```

As you can see one side of the comparisons seems to think this is `['b']` instead of `['a', 'b']` - it's like using multiple values replaces the previous values with the latest.

Using the `parse` command:

```
sqlfluff parse --dialect=bigquery test/fixtures/parser/bigquery/select_mixture_of_array_literals.sql -f yaml
```

This does handle this fine. So it's just a test suite issue.

Trying it with a `-` (I [never really understood this in YAML](https://www.reddit.com/r/ansible/comments/5jhff3/when_to_use_dash_in_yaml/) to be honest!)

```yml
            array_literal:
              - start_square_bracket: '['
              - expression:
                literal: "'a'"
              - comma: ','
              - expression:
                literal: "'b'"
              - end_square_bracket: ']'
```

Gives a slightly different error with a clue what's wrong:

```py
    def process_struct(obj):
        """Process a nested dict or dict-like into a check tuple."""
        if isinstance(obj, dict):
            return tuple((k, process_struct(obj[k])) for k in obj)
        elif isinstance(obj, list):
            # We'll assume that it's a list of dicts
            if isinstance(obj[0], dict):
                buff = [process_struct(elem) for elem in obj]
                if any(len(elem) > 1 for elem in buff):
>                   raise ValueError(f"Not sure how to deal with multi key dict: {buff!r}")
E                   ValueError: Not sure how to deal with multi key dict: [(('start_square_bracket', '['),), (('expression', None), ('literal', "'a'")), (('comma', ','),), (('expression', None), ('literal', "'b'")), (('end_square_bracket', ']'),)]

test/conftest.py:28: ValueError
```

So it's not a big issue, in that the code still works with multiple elements in real life. But it does limit our test suite ability, and can be frustrating when getting weird random errors when you're trying to do the right thing and add good test cases.

This looks like a hard core Python and SQLFluff so not sure I have the skills to help with this one (I only dabble in Python), so not sure if one of the maintainers can comment on it?

**Comment by barrywhart** [comment]
In cases like the above, what does the `sqlfluff parse` output look like? Is it also missing those items? Just a quick question, trying to maybe narrow down where the problem may be.

**Comment by tunetheweb** [comment]
Mentioned that in the issue. The parse command is fine - all the terms are there as expected.

**Comment by barrywhart** [comment]
Got it -- sorry, I was skimming and missed it. Thanks for the detailed report! If I have time, I'll try and look at this in the next week or two.

**Comment by barrywhart** [comment]
@tunetheweb: I see the issue now in your description. The interesting thing is, I'm seeing different behavior on my machine. See the PR I created (#1275) and notice the hyphens in the updated YAML. I didn't add those -- the `generate_parse_fixture_yml.py` script added them automatically.

Do you remember my PR last week when you pointed out that some YAML files had changed for no reason? I think there may be a bug either in our script or in the underlying YAML package resulting in inconsistent YAML output. I notice that our `requirements` files do not pin the versions of underlying packages. Can you check and tell me what versions of these packages you are using?

* `oyaml`
* `PyYAML`

On my machine, I have:
* `oyaml==1.0`
* `PyYAML==5.4.1`

I may look into pinning the versions of these and other packages -- this will help avoid variation.

**Comment by barrywhart** [comment]
And... now I have another theory. This [section](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/parser/segments/base.py#L398L406) of code is an intermediate step when converting parse results to YAML files. It's trying to be clever about whether to create a dictionary or list. TBH, I don't _see_ any issues with the code, but there's a "smell" here: if this code creates a dictionary, the resulting YAML won't have hyphens, but if it creates a list, it will. That's very close to what we're seeing. Maybe there's a corner case where it makes the wrong choice sometimes, perhaps a bit arbitrarily?

In my experimental PR for this issue (#1275), I replaced this block of code with a single line:
```
                elem = [cls.structural_simplify(e) for e in elem]
```

This changes most of the YAML files, making them harder to read. So I'm not recommending this change, just trying to understand the behavior.

**Comment by tunetheweb** [comment]
I've got those versions too and checking out your branch worked for me now.

I fear I may have led your on wild goose chase here as I was editing the YML files by hand for speed. And now I can't repeat this when doing them automatically! Will try again over the weekend with a fresh head but at the moment it looks like this was user error 😞

**Comment by barrywhart** [comment]
No worries, it was interesting looking at that area of the code. It's still a mystery to me why the YAML files sometimes change. Maybe other people are editing them by hand as well. 🤔

**Comment by tunetheweb** [comment]
Yup confirmed it as me not knowing how to generate the YAML files. Huge apologies for wasting your time.

I've added a link to the CONTRIBUTING.md file in #1276 along with other clean ups discussed in this issue.
